source/extensions/tracers/xray/daemon_broker.cc

```diff
@@ -0,0 +1,58 @@
+#include "extensions/tracers/xray/daemon_broker.h"
+
+#include "common/network/utility.h"
+
+#include "source/extensions/tracers/xray/daemon.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace XRay {
+
+namespace {
+// creates a header JSON for X-Ray daemon.
+// For example:
+// { "format": "json", "version": 1}
+std::string createHeader(const std::string& format, uint32_t version) {
+  using namespace envoy::tracers::xray;
+  daemon::Header header;
+  header.set_format(format);
+  header.set_version(version);
+
+  Protobuf::util::JsonPrintOptions json_options;
+  json_options.preserve_proto_field_names = true;
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(header, &json, json_options);
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize X-Ray header to JSON.");
+  }
+  return json;
+}
+
+} // namespace
+
+DaemonBrokerImpl::DaemonBrokerImpl(const std::string& daemon_endpoint) {
+  using Network::Utility;
+  auto& logger = Logger::Registry::getLog(Logger::Id::tracing);
+
+  const auto address = Utility::parseInternetAddressAndPort(daemon_endpoint, false /*v6only*/);
+  io_handle_ = address->socket(Network::Address::SocketType::Datagram);
+  if (io_handle_->fd() == -1) {
+    ENVOY_LOG_TO_LOGGER(logger, error, "Failed to acquire UDP socket to X-Ray daemon at - {}",
+                        daemon_endpoint);
```

Comment:
good catch. fixed.

Definitely not throwing :) I dislike exceptions so much.

Reply:
save status in the object then?

Reply:
Can you explain your thoughts about adding a status data member in this class?

I'm honestly not sure when it would be the case of failing to acquire a UDP socket, and whether that is something that would persist or might change if we keep retrying...

Reply:
Yes, if something goes wrong opening the socket, we need to define what the system behavior should be after that.

I'm not sure whether the behavior should be:
 * crash with error message
 * run in degraded state
 * something else

But I don't have a clear idea on what the behavior should be in this case.

Reply:
Well, right now the behavior is to log the error and retry again the next time.. I think that is acceptable personally, it can also be thought of as running in a degraded state as well. 

Is there a particular reason you are _not_ OK with that approach?

Reply:
Just paranoia. What happens if it's in a state it never succeeds? Will it spam the log file?

I'm OK with your approach if you just add some comments saying what happens & why that's OK.

Reply:
Acquiring the socket itself should never fail IIRC so this can be a release assert. The connect below can definitely fail, and that should not be an error case, as it can spam the log. I would recommend actually getting rid of connect entirely and just using sendmsg() directly (see all of the UDP code that has been merged recently). This will avoid an issue like: https://github.com/envoyproxy/envoy/issues/8911 which I'm going to fix this week by removing the connect and moving to sendmsg() :)

Reply:
Looking at `Network::IoHandle::sendmsg(...)`, is there a way to create a `RawSlice` without copying the data into one of the `Slice` abstractions? seems like a bit of an overkill compared to the current solution (modulo the log spam problem).

Reply:
https://github.com/envoyproxy/envoy/blob/80cea119e0910e9c41ba543017bfc3ee0cb0909f/source/common/network/utility.h#L303-L305

Reply:
I guess I'd have to ask the same question about `Buffer::Instance`. Is there a way to create one without copying the data into it?

Reply:
Just use the slice variant and point to the data. It's just a string view. See other examples in the code base.